### DISCUSSION TITLE
Don't try to autoload relations when reporting errors!

### DISCUSSION THREAD

**Discussion by dtkagi** [sqlalchemy/sqlalchemy#11674] [Usage Questions] [answered]
I wasted way too much time trying to decipher what errors such as this mean:
>                | sqlalchemy.orm.exc.DetachedInstanceError: Parent instance <Source at 0xffff6219ffe0> is not bound to a Session; lazy load operation of attribute 'articles' cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)

For example issue was simple and I was just passing: function_that_want_ids(source) and then error happens.

I was questioning my session management code instead because in async environment error is poor and why does it fails when relations aren't loaded in first place? Why don't you simply check this before generating errors?

I am quite experienced programmer that codes in python from time to time and this hit me 3rd time in a row. So annoying.

**Comment by zzzeek** [comment]
> I wasted way too much time trying to decipher what errors such as this mean:

did you read the documentation linked in the error message? 

> 
> I was questioning my session management code instead because in async environment error is poor and why does it fails when relations aren't loaded in first place? Why don't you simply check this before generating errors?

SQLAlchemy defaults to lazy loading for relationships, what would be "checked" ?   as the document indicates, use eager loading for those relationships that you want to be available for detached objects

**Reply by dtkagi** [reply]
@zzzeek Yes I've read the documentation. 

The problem is that when some error happens the object tries to serialize these "lazy" fields (for error stacktrace needs) and fails with error above.

Simplest reproduction scenario is like this:
```
entity = find()
print(entity)
```
Gonna throw this error for me. It could be due to some other dependency maybe because I can't believe it happens for more people and wasn't addressed.

Additional context:
> poetry show | grep sql
sqlalchemy                2.0.25          Database Abstraction Library


After updating to 2.0.31  I am getting https://docs.sqlalchemy.org/en/20/errors.html#error-xd2s instead. So it still tries to load some relations when I `print(entity)`

**Reply by dtkagi** [reply]
I suspect that you simply don't filter lazy attributes before serialization.

Ok, when I change code to:
```py
class Source(Base):
    __tablename__ = "source"

    def __repr__(self):
        return f"Now it doesn't fail"
```
Problem is gone.

Maybe I use some niche/incorrect Base class?
```py
from sqlalchemy.orm import (
    DeclarativeBase,
    Mapped,
    MappedAsDataclass,
    mapped_column,
    relationship,
)

class Base(MappedAsDataclass, DeclarativeBase):
    pass
    
class Source(Base):
    __tablename__ = "source"

    articles: Mapped[List["Article"]] = relationship(
        "Article", back_populates="source", init=False
    )
    
    id: Mapped[int] = mapped_column(
        init=True, primary_key=True, autoincrement=True, default=None
    )
    
    
 x = Source()
 print(x)
```

This is minimal reproduction scenario

**Reply by CaselIT** [reply, marked as answer]
You have definitely the models as a dataclass. You need to look at the dataclass documentation on how to control it repr. By default it will log every field in the repr if I'm not mistaken.


Also what's the error that you are talking about that logs the object?

**Reply by dtkagi** [reply]
> You have definitely the models as a dataclass. You need to look at the dataclass documentation on how to control it repr

Thank you, that's most likely it.

BTW do I need to use this `MappedAsDataclass`? I think I took it from example from documentation and not actually sure. Ok from quick check it seems it's needed to use `init`. Is there a chance to patch `MappedAsDataclass` export to fix repr? 


> Also what's the error that you are talking about that logs the object?

When any error happens this serialization kicks in (to display error) and it's confusing because you expect errors where you change stuff and then the stacktrace is ultra cryptic and it comes down to this `__repr__` issue.

**Reply by gordthompson** [reply]
Just declare your relationship with `init=False, repr=False`
